feat(Teacher): Implement unit tagging

geoffreykwan opened the pull request:
## Changes
- Added the ability to create, edit, and delete tags
- Added the ability to tag units in the run list and personal unit library
- Added the ability to filter units by tag in the run list and personal unit library

## Test
- Test with https://github.com/WISE-Community/WISE-API/pull/276
- Make sure all the tagging functionality works as expected

Closes #1686

geoffreykwan committed:
feat(Teacher): Implement apply tags button (#1687)

geoffreykwan committed:
feat(Tag): Implement manage tags dialog (#1693)

geoffreykwan committed:
feat(Tag): Implement create tag dialog

geoffreykwan committed:
feat(Tag): Show snack bar message when a tag is created

geoffreykwan committed:
refactor(Tag): Create project tag service

geoffreykwan committed:
Merge pull request #1707 from WISE-Community/add-new-tag-dialog

feat(Tag): Implement create tag dialog

geoffreykwan committed:
Merge pull request #1715 from WISE-Community/create-project-tag-service

refactor(Tag): Create project tag service

geoffreykwan committed:
feat(Tag): Use tag objects instead of strings in project

geoffreykwan committed:
feat(Tag): Sort tags in apply tag and run item lists when tag changes

geoffreykwan committed:
chore(Tag): Add private modifier to function

geoffreykwan committed:
Merge pull request #1716 from WISE-Community/use-tag-objects-in-project

feat(Tag): Use tag objects instead of strings in project

geoffreykwan committed:
feat(Tag): Implement delete tag

geoffreykwan committed:
refactor(Tag): Clean up some code

geoffreykwan committed:
Merge pull request #1718 from WISE-Community/delete-project-tag

feat(Tag): Implement delete tag

geoffreykwan committed:
feat(Tag): Prevent creating a duplicate tag

geoffreykwan committed:
refactor(Tag): Clean up code and add private function modifier

geoffreykwan committed:
Merge pull request #1722 from WISE-Community/prevent-creating-duplicate-tags

feat(Tag): Prevent creating a duplicate tag

geoffreykwan committed:
feat(Tag): Prevent duplicate tags when editing tag

geoffreykwan committed:
refactor(Tag): Remove unused code

geoffreykwan committed:
refactor(Tag): Remove unused providers

geoffreykwan committed:
refactor(Tag): Clean up code

geoffreykwan committed:
Merge pull request #1723 from WISE-Community/prevent-editing-duplicate-tags

feat(Tag): Prevent duplicate tags when editing tag

geoffreykwan committed:
feat(Tag): Implement tag color

geoffreykwan committed:
Merge pull request #1727 from WISE-Community/implement-tag-color

feat(Tag): Implement tag color

geoffreykwan committed:
feat(Tag): Hide archived tag and prevent user from creating archived tag

geoffreykwan committed:
feat(Tag): Filter out archived tag when retrieving tags

geoffreykwan committed:
Merge pull request #1730 from WISE-Community/hide-archived-tag

feat(Tag): Hide archived tag and prevent user from creating archived tag

geoffreykwan committed:
feat(Tag): Use inline editing instead of popup dialogs to edit tags

geoffreykwan committed:
fix(Tag): Show indeterminate checkbox when only some projects have a tag

geoffreykwan committed:
refactor(Tag): Reduce the number of times ngOnChanges() is called

geoffreykwan committed:
feat(Tag): Clean up some code

geoffreykwan committed:
Merge pull request #1735 from WISE-Community/tag-inline-editing

feat(Tag): Use inline editing instead of popup dialogs to edit tags

geoffreykwan committed:
refactor(Tag): Clean up some code

geoffreykwan committed:
refactor(Tag): Use SelectAllItemsCheckboxComponent for applying tags

geoffreykwan committed:
refactor(Tag): Convert SelectAllItemsCheckboxComponent to standalone

geoffreykwan committed:
Merge pull request #1736 from WISE-Community/fix-apply-tag-checkbox-indeterminate

fix(Tag): Show indeterminate checkbox when only some projects have a tag

geoffreykwan committed:
feat(Tag): Add color chooser when editing tag

geoffreykwan committed:
test(Tag): Fix tests

geoffreykwan committed:
fix(Tag): Fix new tag checked value

geoffreykwan committed:
Merge pull request #1743 from WISE-Community/fix-new-tag-checked-value

fix(Tag): Fix new tag checked value

geoffreykwan committed:
feat(Tag): Create color chooser component

geoffreykwan committed:
feat(Tag): When editing tag show preview with text and color

geoffreykwan committed:
feat(Tag): Color chooser now sends a color string

geoffreykwan committed:
Merge pull request #1742 from WISE-Community/tag-color-chooser

feat(Tag): Add color chooser when editing tag

geoffreykwan committed:
feat(Tag): Add search input to the apply tags menu

geoffreykwan committed:
feat(Tag): Clear search text when apply tags menu is closed

geoffreykwan committed:
feat(Tag): Clear search text in template when menu is closed

geoffreykwan committed:
Merge pull request #1750 from WISE-Community/tag-search

feat(Tag): Add search input to the apply tags menu

geoffreykwan committed:
refactor(Tag): Convert apply tags button to standalone

geoffreykwan committed:
Merge pull request #1757 from WISE-Community/convert-apply-tags-button-to-standalone

refactor(Tag): Convert apply tags button to standalone

geoffreykwan committed:
feat(Tag): Add filter units by tag

geoffreykwan committed:
refactor(Authoring): Rename TagsFilter to SelectTags

Also rename tagsFilterChanged to selectTagEvent.

geoffreykwan committed:
refactor(Tag): Rename a variable and function

geoffreykwan committed:
Merge pull request #1761 from WISE-Community/add-filter-units-by-tag

feat(Tag): Add filter units by tag

geoffreykwan committed:
feat(Tag): Show selected tags list

geoffreykwan committed:
feat(Tag): Use mat chip for run card tags

breity committed:
Add keyboard accessibility for remove filter buttons.
Update mat-chip label font-size to match buttons.

breity committed:
Add aria-label to tag list

geoffreykwan committed:
feat(Tag): Clear tags using the existing clear filters link

Also fix message when there are no active runs or no archived runs.

geoffreykwan committed:
Merge pull request #1775 from WISE-Community/use-mat-chip-for-run-card-tags

feat(Tag): Use mat chip for run card tags

geoffreykwan committed:
Merge branch 'issue-1686-implement-unit-tagging' into show-selected-tags-list

geoffreykwan committed:
Merge pull request #1774 from WISE-Community/show-selected-tags-list

feat(Tag): Show selected tags list

geoffreykwan committed:
feat(Tag): Add tag functionality to personal library

geoffreykwan committed:
feat(Tag): Add optional clear filters button

breity committed:
Fix tags filter alignment and update styles

geoffreykwan committed:
fix(Tag): Fix apply tags checkbox values not updating in library

geoffreykwan committed:
Merge pull request #1782 from WISE-Community/add-tag-functionality-to-personal-library

feat(Tag): Add tag functionality to personal library

geoffreykwan committed:
Merge branch 'develop' into issue-1686-implement-unit-tagging

geoffreykwan committed:
test(Tag): Fix tests

hirokiterashima commented on src/app/teacher/select-tags/select-tags.component.ts:
```diff
@@ -0,0 +1,19 @@
+  templateUrl: './select-tags.component.html',
+  styleUrls: ['./select-tags.component.scss'],
+  standalone: true,
+  imports: [CommonModule, MatDividerModule, SearchBarComponent, SharedModule]
```
Should we import SharedModule here? It includes things that we don't use like UnlinkGoogleAccountConfirmComponent and UnlinkGoogleAccountPasswordComponent. Instead, can we import only the classes that we need?

hirokiterashima commented on src/app/teacher/apply-tags-button/apply-tags-button.component.ts:
```diff
@@ -0,0 +1,78 @@
+    MatTooltipModule,
+    SearchBarComponent,
+    SelectAllItemsCheckboxComponent,
+    SharedModule
```
Should we import SharedModule here? It includes things that we don't use like UnlinkGoogleAccountConfirmComponent and UnlinkGoogleAccountPasswordComponent. Instead, can we import only the classes that we need?

hirokiterashima commented on src/app/teacher/teacher.module.ts:
```diff
@@ -43,12 +43,18 @@ import { MatListModule } from '@angular/material/list';
   MatButtonModule,
   MatCardModule,
   MatCheckboxModule,
+  MatChipsModule,
```
Is this import necessary?

hirokiterashima commented on src/app/modules/library/select-all-items-checkbox/select-all-items-checkbox.component.spec.ts:
```diff
@@ -9,8 +9,7 @@ describe('SelectAllItemsCheckboxComponent', () => {
     TestBed.configureTestingModule({
-      declarations: [SelectAllItemsCheckboxComponent],
-      imports: [MatCheckboxModule, MatTooltipModule]
+      imports: [MatCheckboxModule, MatTooltipModule, SelectAllItemsCheckboxComponent]
```
Are MatCheckboxModule and MatTooltipModule needed?

hirokiterashima commented on src/app/teacher/color-chooser/color-chooser.component.spec.ts:
```diff
@@ -0,0 +1,22 @@
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      imports: [ColorChooserComponent, CommonModule, MatChipsModule]
```
Are CommonModule and MatChipsModule needed?

hirokiterashima commented on src/app/teacher/edit-tag/edit-tag.component.spec.ts:
```diff
@@ -0,0 +1,30 @@
+        BrowserAnimationsModule,
+        EditTagComponent,
+        HttpClientTestingModule,
+        MatSnackBarModule
```
This this import needed?

hirokiterashima commented on src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.spec.ts:
```diff
@@ -0,0 +1,31 @@
+      imports: [
+        HttpClientTestingModule,
+        ManageTagsDialogComponent,
+        MatSnackBarModule,
```
Is this import needed?

hirokiterashima commented on src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts:
```diff
@@ -0,0 +1,93 @@
+  constructor(
+    private snackBar: MatSnackBar,
+    private teacherService: TeacherService,
+    private projectTagService: ProjectTagService
```
alphabetize?

hirokiterashima reviewed (changes requested):
The functionality mostly works. 

It looks like the number of tags in the message when you try to delete a tag is off. In the screenshot below, you can see that I have "tag3" on a unit, but the cofirmation message says "0" I think it may not be counting the tags that are on the units (as opposed to run units).

<img width="300" alt="Screenshot 2024-05-24 at 3 30 00 PM" src="https://github.com/WISE-Community/WISE-Client/assets/119416/f2355334-f845-4d31-9ba6-b1f66459debd">


I also added some code cleaning suggestions inline. 

Lastly, maybe we should also apply the code cleaning conventions for at least the new standalone components, like alphabetizing (imports -> selector -> standalone...), rename "styleUrls" -> "styleUrl", self-closing tags, and such?

geoffreykwan committed:
fix(Tag): Number of units with tag delete message

geoffreykwan committed:
refactor(Tag): Remove unnecessary imports

geoffreykwan committed:
refactor(Tag): Clean up component parameters and use self closing tags

geoffreykwan commented:
I've fixed the bug and made all the changes.

Make sure to get the latest from the WISE-API branch issue-259-implement-unit-tagging for the fix to work.

geoffreykwan committed:
Merge branch 'develop' into issue-1686-implement-unit-tagging

breity committed:
feat(Tag): Replace mat-chips with UnitTagsComponent, TagComponent
- Update manage and edit tag styles
- Use colorjs library for accessible tag text colors
- Fix padding for dialog content with scroll

hirokiterashima reviewed (changes requested):
When I delete a tag in my unit library, the tag still appears on the unit. I have to refresh the page to see the change. Can they be immediately removed without having to refresh?

The tag choices in the "Apply Tags" filter dialog doesn't seem to update immediately after you add or delete tags. It only seems to update when you edit the tag name or color.

breity committed:
feat(Tag): Use ng-select to select/display unit tags

geoffreykwan committed:
fix(Tag): Fix tags not updating in units and tags list in library

geoffreykwan commented:
I've fixed the tags not updating bugs.

hirokiterashima commented:
I verified the new commit fixes the tags not updating bugs.

The unit tests are failing now. I see a lot of ```NullInjectorError: No provider for ProjectTagService!``` errors.

For all the new html files that were introduced in this PR, (manage-tags-dialog.component.html, color-chooser.component.html, etc) should we apply the same improvements that we've been making in our standalone component work? I think doing this will make the template easier to read. Mainly:
- use new control flow syntax (```@if, @else if, @for,...```)
- use self-closing tags

geoffreykwan committed:
test(Tag): Fix failing tests by adding ProjectTagService provider

geoffreykwan committed:
Merge branch 'develop' into issue-1686-implement-unit-tagging

geoffreykwan committed:
refactor(Tag): Use self closing tags

geoffreykwan committed:
refactor(Tag): Use new template flow control

geoffreykwan committed:
chore(i18n): Generate translations

geoffreykwan commented:
Ok, I've fixed the tests, used the new flow control, and used self closing tags.

hirokiterashima commented on src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.html:
```diff
@@ -0,0 +1,60 @@
+<h2 mat-dialog-title i18n>Manage Tags</h2>
+<mat-dialog-content class="dialog-content-scroll">
+  <div fxLayout="column" fxLayoutGap="12px">
+    @if (!showCreateTag) {
+      <div>
+        <button mat-flat-button color="primary" (click)="showCreateTag = !showCreateTag" i18n>
+          + New Tag
+        </button>
+      </div>
+    }
+    @if (showCreateTag) {
+      <edit-tag (closeEvent)="showCreateTag = false" />
+    }
```
Switch test order and use ```@else``` for better readability?
```suggestion
    @if (showCreateTag) {
      <edit-tag (closeEvent)="showCreateTag = false" />
    } @else {
      <div>
        <button mat-flat-button color="primary" (click)="showCreateTag = !showCreateTag" i18n>
          + New Tag
        </button>
      </div>
    }
```

hirokiterashima commented on src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.html:
```diff
@@ -0,0 +1,60 @@
+                <div class="edit-tag">
+                  <edit-tag [tag]="tag" (closeEvent)="cancelEditing(tag)" />
+                </div>
+              }
+              @if (!idToEditing[tag.id]) {
```
Use ```@else``` here?
```suggestion
              } @else {
```

hirokiterashima reviewed (commented):
New changes look good. I added some more code suggestions inline to improve control flow readability.

Also, I noticed that in the "Apply Tags" select dropdown, clicking on the choice span (not the checkbox, not the tag text, but the space next to those on the same row) does not add/delete the tag. The cursor is a pointer, so I expected it to add/delete. We should handle clicking on the entire row?

geoffreykwan committed:
refactor(Tag): Use @if/@else control flow

geoffreykwan committed:
fix(Tag): Make apply tags options take up the whole width

geoffreykwan commented:
I've changed the flow control to use @if/@else and fixed the apply tags option bug.

breity committed:
feat(Tag): Use TagComponent in apply tags dropdown

breity committed:
feat(tag): Add checkboxes to SelectTagsComponent ng-select
- Clean up code

geoffreykwan committed:
refactor(Tag): Clean up code

breity committed:
Fix tag filter dropdown max width

breity committed:
feat(Tag): Show unit tags in project details dialog

breity committed:
Add margin-bottom to unit tags in project details dialog

geoffreykwan committed:
feat(Tag): Show applied and removed tag snack bar message

hirokiterashima reviewed (approved):
LGTM

breity reviewed (approved):
Looks good. 👍

geoffreykwan commented:
:tada: This PR is included in version 5.154.0 :tada:

The release is available on [GitHub release](https://github.com/WISE-Community/WISE-Client/releases/tag/v5.154.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
